### PR TITLE
15 as a user i want to mint shared metadata nfts into a collection

### LINKED ISSUE #15: As a User I want to mint shared metadata NFTs into a collection
As a User, I want to mint shared metadata NFTs into a collection.

**Acceptance Criteria:**
- There is a way to mint shared metadata NFTs providing the `tokenId`, `quantity`, and `metadata CID`. 
- There should be no limit on how many NFTs can be minted in a single function call.
- Function should return `List<NftId>` created.
- The function does not estimate the total cost of all the transactions before execution.
- If the function fails, return the `List<NftId>` that was successfully minted before the error was thrown.
- There should be a parameter that lets the user specify how many NFTs should be minted in a single batch. The values allowed should be integers from 1 to 10. The default value should be 5.

### PR DESCRIPTION
**Pull request by StanislawMazowieckiAriane**
**Related issue(s)**:

Fixes #15 

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: src/dictionary/errors.json [hunk 1]
```diff
@@ -5,5 +5,12 @@
   "collectionSymbolRequired": "collectionSymbol is required",
   "myAccountIdRequired": "myAccountId is required",
   "treasuryAccountPrivateKeySignRequired": "If you want to use treasuryAccount to sign, you need to pass the treasuryAccountPrivateKey also",
-  "collectionNotCreated": "Something went wrong while creating the collection"
+  "collectionNotCreated": "Something went wrong while creating the collection",
+  "mintingError": "There was an error while minting the NFT. Successfully minted: ",
+  "maxBuffer": "Max Buffer exceeded. Use buffer smaller than 10",
```
Comment (BartoszSolkaArianelabs) [inline src/dictionary/errors.json:10]: smaller or equal to 10

File: src/functions/mintSharedMetadataFunction.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  buffer,
+  metaData,
+  supplyKey,
+}: MintTokenType) => {
+  if (buffer > 10) throw new Error(errors.maxBuffer);
+  if (buffer < 1) throw new Error(errors.minBuffer);
+  if (!tokenId) throw new Error(errors.tokenIdRequired);
+  if (!amount) throw new Error(errors.minAmount);
+  if (!metaData) throw new Error(errors.metadataRequired);
+  if (!supplyKey) throw new Error(errors.supplyKeyRequired);
```
Comment (BartoszSolkaArianelabs) [inline src/functions/mintSharedMetadataFunction.ts:18]: supply key should be optional.

In the simplest token, a supply key is the same as the client key

File: test/unit/mintSharedMetadataFunction.test.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        metaData: mockMetaData,
+        tokenId: mockTokenId,
+        supplyKey: mockSupplyKey,
+        amount: mockAmount,
+        buffer: mockBuffer,
+      })
+    ).rejects.toThrow(errors.minBuffer);
+  });
+
+  it('should handle error when buffer is higher than 10', async () => {
```
Comment (BartoszSolkaArianelabs) [inline test/unit/mintSharedMetadataFunction.test.ts:99]: what is the difference between this one and `should handle error when buffer is bigger than 10`?

File: src/functions/mintSharedMetadataFunction.ts (hunk 2 above)
Comment (se7enarianelabs) [inline src/functions/mintSharedMetadataFunction.ts:18]: This could be extracted to a separate function as the same validation will probably be in a ` mintUniqueMetadataFunction` also the code will be cleaner.

File: src/functions/mintSharedMetadataFunction.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  const successMetadata = [];
+  const numberOfCalls = Math.ceil(amount / buffer);
```
Comment (se7enarianelabs) [inline src/functions/mintSharedMetadataFunction.ts:21]: Add comment, explaining the formula

File: src/functions/mintSharedMetadataFunction.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import { MintTokenType } from '../types/mintToken';
+import { tokenMinter } from './tokenMinter';
+import errors from '../dictionary/errors.json';
+
+export const mintSharedMetadataFunction = async ({
+  client,
+  tokenId,
+  amount,
+  buffer,
```
Comment (se7enarianelabs) [inline src/functions/mintSharedMetadataFunction.ts:9]: This should be renamed to batchSize

File: src/functions/mintSharedMetadataFunction.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  metaData,
+  supplyKey,
+}: MintTokenType) => {
+  if (buffer > 10) throw new Error(errors.maxBuffer);
```
Comment (se7enarianelabs) [inline src/functions/mintSharedMetadataFunction.ts:13]: Should be >=

File: src/functions/mintSharedMetadataFunction.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  if (buffer < 1) throw new Error(errors.minBuffer);
```
Comment (se7enarianelabs) [inline src/functions/mintSharedMetadataFunction.ts:14]: should be <= 1

File: src/functions/mintSharedMetadataFunction.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  if (!amount) throw new Error(errors.minAmount);
+  if (!metaData) throw new Error(errors.metadataRequired);
+  if (!supplyKey) throw new Error(errors.supplyKeyRequired);
+
+  const successMetadata = [];
+  const numberOfCalls = Math.ceil(amount / buffer);
+
+  try {
+    for (let i = 0; i < numberOfCalls; i++) {
+      const optionsHelper = new Array(Math.min(buffer, amount)).fill(metaData);
```
Comment (se7enarianelabs) [inline src/functions/mintSharedMetadataFunction.ts:25]: Can it be renamed to something more descriptive?

File: src/functions/mintSharedMetadataFunction.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      amount -= buffer;
+      await tokenMinter(optionsHelper, tokenId, supplyKey, client);
+      successMetadata.push(optionsHelper);
+    }
+
+    return successMetadata.flat(1);
```
Comment (se7enarianelabs) [inline src/functions/mintSharedMetadataFunction.ts:31]: isn't default value for flat `1`?

File: src/functions/mintSharedMetadataFunction.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  } catch (error) {
+    throw new Error(`${errors.mintingError} ${successMetadata.flat(1).join(' ')}`);
```
Comment (se7enarianelabs) [inline src/functions/mintSharedMetadataFunction.ts:33]: isn't default value for flat `1`?

File: src/functions/tokenMinter.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import { Client, Hbar, PrivateKey, TokenMintTransaction } from '@hashgraph/sdk';
+
+export async function tokenMinter(
+  metaData: string[],
+  tokenId: string,
+  supplyKey: string,
```
Comment (se7enarianelabs) [inline src/functions/tokenMinter.ts:6]: Why isn't it a PrivateKey type?

File: src/functions/tokenMinter.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  client: Client
+) {
+  const CIDs = metaData.map((metaData) => Buffer.from(metaData));
+
+  const transaction = new TokenMintTransaction()
+    .setTokenId(tokenId)
+    .setMaxTransactionFee(new Hbar(20))
```
Comment (se7enarianelabs) [inline src/functions/tokenMinter.ts:13]: `new Hbar(20)` is a magic number, maybe create a const

File: src/index.ts [hunk 13]
```diff
@@ -1,21 +1,28 @@
...
 import { HederaNFTSDK } from './HederaNFTSDK';
+import 'dotenv/config';
 
 const myPrivateKey = process.env.FIRST_PRIVATE_KEY!;
 const myAccountId = process.env.FIRST_ACCOUNT_ID!;
 
 const nftSDK = new HederaNFTSDK(myAccountId, myPrivateKey);
 
 (async () => {
-  const tokenId = await nftSDK.createCollection('test', 'test2');
```
Comment (se7enarianelabs) [inline src/index.ts:11]: What is this file for?

File: src/index.ts [hunk 14]
```diff
@@ -1,21 +1,28 @@
...
+  // const tokenId = await nftSDK.createCollection('test', 'test2');
```
Comment (se7enarianelabs) [inline src/index.ts:11]: Why is this line commented?

File: src/types/mintToken.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+import { Client } from '@hashgraph/sdk';
+
+export type MintTokenType = {
+  client: Client;
+  tokenId: string;
+  amount: number;
+  buffer: number;
```
Comment (se7enarianelabs) [inline src/types/mintToken.ts:7]: maybe rename to batchSize?

File: test/unit/mintSharedMetadataFunction.test.ts [hunk 16]
```diff
@@ -0,0 +1,186 @@
+import { mintSharedMetadataFunction } from '../../src/functions/mintSharedMetadataFunction';
+import { tokenMinter } from '../../src/functions/tokenMinter';
+import { Client } from '@hashgraph/sdk';
+import errors from '../../src/dictionary/errors.json';
+
+jest.mock('../../src/functions/tokenMinter', () => ({
+  tokenMinter: jest.fn(),
+}));
+
+describe('mintUniqueMetadataFunction', () => {
```
Comment (se7enarianelabs) [inline test/unit/mintSharedMetadataFunction.test.ts:10]: typo, mintShared

File: test/unit/mintSharedMetadataFunction.test.ts [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+      tokenId: mockTokenId,
+      supplyKey: mockSupplyKey,
+      amount: mockAmount,
+      buffer: mockBuffer,
+    });
+
+    expect(result).toEqual(expect.arrayContaining([mockMetaData]));
+  });
+
+  it('should handle amount less than buffer correctly', async () => {
```
Comment (se7enarianelabs) [inline test/unit/mintSharedMetadataFunction.test.ts:33]: You should also test the edge values for Buffer so 1 and 10

File: src/functions/mintSharedMetadataFunction.ts (hunk 7 above)
Comment (StanislawMazowieckiAriane) [reply src/functions/mintSharedMetadataFunction.ts:14]: Are you sure that the buffer can't be equal to 1?

File: src/index.ts (hunk 13 above)
Comment (StanislawMazowieckiAriane) [reply src/index.ts:11]: Now for tests in dev.
Later there would be export

File: src/functions/mintSharedMetadataFunction.ts (hunk 7 above)
Comment (se7enarianelabs) [reply src/functions/mintSharedMetadataFunction.ts:14]: You are right, my mistake

File: src/index.ts (hunk 13 above)
Comment (se7enarianelabs) [reply src/index.ts:11]: I don't think this file should be committed, what do you think @BartoszSolkaArianelabs

File: src/index.ts (hunk 14 above)
Comment (BartoszSolkaArianelabs) [reply src/index.ts:11]: IMO index.ts should not be the place to test the code. We have end2end tests for it

File: src/index.ts (hunk 13 above)
Comment (BartoszSolkaArianelabs) [reply src/index.ts:11]: IMO index.ts should not be the place to test the code. We have end2end tests for it

File: test/unit/mintSharedMetadataFunction.test.ts (hunk 3 above)
Comment (StanislawMazowieckiAriane) [reply test/unit/mintSharedMetadataFunction.test.ts:77]: done

File: test/e2e/mintSharedMetadataFunctionE2E.test.ts [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+      tokenId,
+      1,
+      2,
+      'www.youtube.com',
+      myPrivateKey
+    );
+
+    expect(tokenId).toBeDefined();
+    expect(transactionStatus).toBeDefined();
+    expect(transactionStatus).toEqual(['www.youtube.com']);
```
Comment (BartoszSolkaBD) [inline test/e2e/mintSharedMetadataFunctionE2E.test.ts:37]: this test should also check if tokens are created on hedera network

File: test/e2e/mintSharedMetadataFunctionE2E.test.ts [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+      tokenId,
+      10,
+      5,
+      'www.youtube.com',
+      myPrivateKey
+    );
+
+    expect(tokenId).toBeDefined();
+    expect(transactionStatus).toBeDefined();
+    expect(transactionStatus).toEqual([
```
Comment (BartoszSolkaBD) [inline test/e2e/mintSharedMetadataFunctionE2E.test.ts:52]: this test should also check if tokens are created on hedera network

File: test/e2e/mintSharedMetadataFunctionE2E.test.ts [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+    const transactionStatus = await nftSDK.mintSharedMetadata(
+      tokenId,
+      3,
+      1,
+      'www.youtube.com',
+      myPrivateKey
+    );
+
+    expect(tokenId).toBeDefined();
+    expect(transactionStatus).toBeDefined();
```
Comment (BartoszSolkaBD) [inline test/e2e/mintSharedMetadataFunctionE2E.test.ts:77]: this test should also check if tokens are created on hedera network

File: test/e2e/mintSharedMetadataFunctionE2E.test.ts [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+    expect(transactionStatus).toEqual(['www.youtube.com', 'www.youtube.com', 'www.youtube.com']);
+  }, 25000);
+
+  it('throws an error when invalid token ID is provided', async () => {
```
Comment (BartoszSolkaBD) [inline test/e2e/mintSharedMetadataFunctionE2E.test.ts:81]: this has a wrong name

Comment (BartoszSolkaBD) [inline test/e2e/mintSharedMetadataFunctionE2E.test.ts:81]: this should be a unit test as it does not interact with any external components. Just checking if the validation works fine

File: test/unit/mintSharedMetadataFunction.test.ts [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+      tokenId: mockTokenId,
+      supplyKey: mockSupplyKey,
+      amount: mockAmount,
+      batchSize: mockBatchSize,
+    });
+
+    expect(result).toEqual(expect.arrayContaining([mockMetaData]));
+  });
+
+  it('should handle error when batchSize is lower than 0', async () => {
```
Comment (BartoszSolkaBD) [inline test/unit/mintSharedMetadataFunction.test.ts:99]: batch size in this test is not lower than 0, it's exactly 0.

File: test/unit/mintSharedMetadataFunction.test.ts [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+        metaData: mockMetaData,
+        tokenId: mockTokenId,
+        supplyKey: mockSupplyKey,
+        amount: mockAmount,
+        batchSize: mockBatchSize,
+      })
+    ).rejects.toThrow(errors.tokenIdRequired);
+  });
+
+  it('should handle error when amount is not passed', async () => {
```
Comment (BartoszSolkaBD) [inline test/unit/mintSharedMetadataFunction.test.ts:187]: amount is passed, it's 0

**Review by BartoszSolkaBD — CHANGES_REQUESTED** [review]

File: test/unit/mintSharedMetadataFunction.test.ts [hunk 24]
```diff
@@ -0,0 +1,208 @@
...
+    const result = await mintSharedMetadataFunction({
+      client: mockClient,
+      metaData: mockMetaData,
+      tokenId: mockTokenId,
+      supplyKey: mockSupplyKey,
+      amount: mockAmount,
+      batchSize: mockBatchSize,
+    });
+
+    expect(result).toEqual(expect.arrayContaining([mockMetaData]));
```
Comment (se7enarianelabs) [inline test/unit/mintSharedMetadataFunction.test.ts:31]: Cannot you monitor how many times tokenMinter was executed?

**Review by se7enarianelabs — CHANGES_REQUESTED** [review]

File: src/functions/mintSharedMetadataFunction.ts [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+      });
+
+      if (result) {
+        mintedMetadata.push(...result);
+      }
+    }
+
+    return mintedMetadata.flat();
+  } catch (error) {
+    throw new Error(`${errors.mintingError} ${mintedMetadata.flat().join(' ')}`);
```
Comment (se7enarianelabs) [inline src/functions/mintSharedMetadataFunction.ts:40]: @StanislawMazowieckiAriane, 
will the developer easily catch this error and retrieve the list of metadata that were omitted? I am not familiar with this syntax: new Error(${errors.mintingError} ${mintedMetadata.flat().join(' ')})?

File: src/functions/mintSharedMetadataFunction.ts [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+  client,
+  tokenId,
+  amount,
+  batchSize,
+  metaData,
+  supplyKey,
+}: MintTokenType) => {
+  validateProps({ tokenId, amount, metaData, supplyKey, buffer: batchSize });
+
+  const mintedMetadata = [];
```
Comment (se7enarianelabs) [inline src/functions/mintSharedMetadataFunction.ts:16]: Rename it, as it no longer contains just the metadata but minted NFT data. Consider renaming it to something like 'mintedNFTs'

**Review by se7enarianelabs — CHANGES_REQUESTED** [review]

File: src/functions/mintSharedMetadataFunction.ts (hunk 25 above)
Comment (StanislawMazowieckiAriane) [reply src/functions/mintSharedMetadataFunction.ts:40]: Yes.
Example: 
Error: There was an error while minting the NFT. Successfully minted:  [{"content":"www.google.com","serialNumber":866},{"content":"www.google.com","serialNumber":867},{"content":"www.google.com","serialNumber":868},{"content":"www.google.com","serialNumber":869},{"content":"www.google.com","serialNumber":870}]

**Review by se7enarianelabs — APPROVED** [review]
